### ISSUE TITLE
Scrollspy / Jump Links (Vertical and Horizontal)

### ISSUE THREAD

**Issue by yih-wang** [patternfly/patternfly-design#792]
In Keycloak, we're applying nav inside a single page that includes multiple cards. Because the page could be very long, and even longer in the future.

The pattern is created with existing components. However, to differentiate with the main navigation, I use different ways to show the selected items - the selected item is just in the blue color, without the short line below it.

I think a visual review would be nice as we don't have such a pattern yet. Thanks.

![image](https://user-images.githubusercontent.com/23331885/72051351-0e3a3880-32fe-11ea-8003-bb4b85685484.png)

**Comment by mcarrano** [comment]
@yih-wang thanks for filing this issue. Are you referring to the nav on the right side of the page?

**Comment by yih-wang** [comment]
> @yih-wang thanks for filing this issue. Are you referring to the nav on the right side of the page?

@mcarrano Right.

**Comment by amysueg** [comment]
@yih-wang I'm meeting with @lboehling about this today and we'll be in touch. There is something like the design you are suggesting in the works on another project.

**Comment by mcarrano** [comment]
@yih-wang It's true that we don't currently have this pattern in PatternFly. But we have been working on something called a progressive form that is similar in that it is a way to place content on a long scrolling form with the ability to navigate to sections. I had created a rough prototype here for testing: https://preview.uxpin.com/cc889a55d206dbb2a21f4c01351624e9c513f212 See the navigation on the left. Would something like this be useful to you? I could possibly see offering this with or without numbered steps.

cc @amysueg

**Comment by amysueg** [comment]
Thanks @mcarrano - I think that it looks like it would translate well to @yih-wang's use case, as long as you're ok moving the nav elements to the left side of the content. Also, I think the visual display will be tighter without the numbers, which seem somewhat unnecessary in this case for sure.

**Comment by yih-wang** [comment]
@mcarrano The prototype looks great. It can perfectly fit my use case without the numbers. My only concern is the position of the navigation. In my use case, users don't need to go through every item of the nav. They may just select some of the items to use. So the navigation has a lower priority in this case. That's why we put the nav in the right. Any thoughts? @amysueg @lboehling Do you have the same use case?

**Comment by amysueg** [comment]
I see what you are saying @yih-wang. Are there any other instances in PF @mcarrano and @lboehling where navigation sits to the right of content? It may be somewhat surprising to see in a UI in which other pages use left to right nav?

**Comment by mcarrano** [comment]
I'm not aware of other instances of this @amysueg . While I understand your reasoning for right-hand placement @yih-wang , I have to admit that on first glance I did not realize that the list on the right was navigation. @kybaker @mceledonia any thoughts here? Alternate layouts you would suggest?

**Comment by mceledonia** [comment]
I've seen some work done with "vertical tabs" similar to what you have here, only it exists on the left hand side and is not in a card, just on the background. I think it would work well here, this is a screenshot:

![image](https://user-images.githubusercontent.com/30935938/72541316-4be41780-3850-11ea-9e1a-062be3e56007.png)

@kybaker Thoughts on bringing that pattern into this product?

**Comment by ajacobs21e** [comment]
The PatternFly Web team has a design in progress for a scrollspy that we plan to contribute to PatternFly Design once the design is complete. @samlinville is working on it - Sam can you give a status update on the design please?

Design mockups: https://sketch.cloud/s/LMyGL
An example in context: https://marvelapp.com/1iiaai96/screen/63401722

We've only considered this from a web perspective so far, so without left navigation. I think this could work in the content area of the page in addition to left nav.


Another idea, could the expandable nav pattern be extended to navigate both within a page and to separate pages?

**Comment by yih-wang** [comment]
Thanks for the suggestions @mceledonia . A reason for it cannot be a tertiary nav is the horizontal tabs in this screen is not the secondary level nav, they are just the sub-tabs of a single object. 

I think this component is more like a scrollspy as @ajacobs21e mentioned. The mockups look great and can perfectly fit our use case. Maybe I can add a hamburger button to the main nav to make it collapsible. So the double nav issue can be relieved to some extent.

**Comment by mcarrano** [comment]
I also like the idea of a Scrollspy nav as a new PatternFly component. It can support this as well as our progressive forms use cases, I think. I updated the title of this issue and will add to our feature backlog. @ajacobs21e and @samlinville, keep me posted on what you have for design and we can discuss contributing to PatternFly.

@LHinson is this something we can get on the roadmap?

**Comment by yih-wang** [comment]
@mcarrano @ajacobs21e Just a quick comment, it will be helpful if there is a mobile display solution when the pattern passed on to PatternFly, as it doesn't have now.

**Comment by ajacobs21e** [comment]
@samlinville is working on finalizing the design, which does have a mobile option. He or I will open a design story when it's ready

**Comment by ajacobs21e** [comment]
@mcarrano 
https://sketch.cloud/s/3WK1w/p/jump-links

Here are the specs for Sam's design. We are proposing the name "Jump links" and this is the vertical variant. PFE has a horizontal sticky nav that we will contribute at some point as "Jump links / Horizontal".

**Comment by ajacobs21e** [comment]
@coreyvickery @djfaucette @samlinville I made a few tweaks to the spec sheet to incorporate both the docs and catalog use cases, most notably the font size of the header and subsections and the width of the highlight border

**Comment by coreyvickery** [comment]
@ajacobs21e Already updated in the kit.

**Comment by yih-wang** [comment]
Hi @mcarrano, Do you know the approximate release time of this component? The Keycloak development is about to start in June/July. So I'm checking the status of the components we're going to use.

**Comment by mcarrano** [comment]
@yih-wang This is currently not scheduled until September/October. It's possible that we can move this up in priority but the will require more discussion. 

cc @LHinson @amysueg

**Comment by mcarrano** [comment]
We should also include a horizontal version of this component as described in the issue here: https://github.com/patternfly/patternfly-design/issues/875

Design:
https://www.sketch.com/s/f6a52c25-8657-4680-9d3e-4c09131e6ed0/a/mYbkyw
https://www.sketch.com/s/f6a52c25-8657-4680-9d3e-4c09131e6ed0/a/Da0Aj2
These are differentiated from tabs by putting the highlight line on top rather than the bottom.
Still need to determine where to put the label.

**Comment by amysueg** [comment]
FYI @dlabaj @srambach

**Comment by mcarrano** [comment]
@lboehling for your reference, here is a vertical jump link implementation that is already available in PatternFly Elements: https://patternflyelements.com/elements/pfe-jump-links/demo/#section2 It would make sense if we can stick relatively close to this approach as long as it satisfies other product use cases. You can reach out to @ajacobs21e if you have questions about this.

**Comment by lboehling** [comment]
Here are the updated designs
Marvel: https://marvelapp.com/project/5200771
Sketch: https://www.sketch.com/s/e6c01dec-8567-436c-b633-eda045100a37/v/WMyrWq

I kept the vertical jump links designed by Allie and Sam the same but added in a second label alignment option for the horizontal jump links. I also included in-context product examples for both vertical and horizontal jump links.

It would be great if we could support options for both left-aligned and centered horizontal jump links as depicted below. My recommendation would be to default to left-aligned but have the ability to center if other elements need to be added in line with the jump links (catalog.redhat.com places logos/link buttons in a band with the horizontal jump links). 

![image](https://user-images.githubusercontent.com/46571616/94197425-c6efda00-fe83-11ea-9202-cc7643d172b0.png)

_________

![image](https://user-images.githubusercontent.com/46571616/94197650-10402980-fe84-11ea-855a-63e6b7a6b5dd.png)
_________

![image](https://user-images.githubusercontent.com/46571616/94197541-f0106a80-fe83-11ea-979d-b778f132c9d9.png)


FYI @ajacobs21e @mcarrano

**Comment by ajacobs21e** [comment]
Love it!! Thanks @lboehling

**Comment by mcarrano** [comment]
This is looking great @lboehling . I just had one minor comment about heading styles that I left in the Marvel deck.

I did want to ask about horizontal jump links as this is something I'm less familiar with. What's the difference between these and tabs? Is it just that the contents of the page scrolls rather than being replaced?  Visually would these just be tabs that have different behaviors associated with them? And if yes, could someone just use the existing horizontal tab component?

**Comment by ajacobs21e** [comment]
Yes that is the difference - navigating to a new section within the same page vs swapping out the content in the section. The design is meant to be complimentary between tabs and jump links but with the highlight line on the opposite side to distinguish between the components.

**Comment by mcarrano** [comment]
I have opened https://github.com/patternfly/patternfly/issues/3558 and https://github.com/patternfly/patternfly/issues/3560 to begin core implementation. Per discussion with @mcoker , this will be split into two parts, the first to create a new Scrollspy component and a second to create a full page demo that shows the scrollspy nav in context.

Thanks for your work on the @lboehling !

**Comment by ajacobs21e** [comment]
horizontal jump links demo from redhat.com https://xd.adobe.com/view/0455e4fb-7d1f-4d15-9b63-1c10fb6979bc-5106/
